stdlib/test/tempfile/test_tempfile.mojo

```diff
@@ -0,0 +1,173 @@
+
+        self.tmpdir_back = os.getenv("TMPDIR")
+        self.temp_back = os.getenv("TEMP")
+        self.tmp_back = os.getenv("TMP")
+
+    fn __enter__(inout self) raises:
+        _ = os.setenv("TMPDIR", self.tmpdir_value, overwrite=True)
+        _ = os.setenv("TEMP", self.temp_value, overwrite=True)
+        _ = os.setenv("TMP", self.tmp_value, overwrite=True)
+
+    fn __exit__(inout self):
+        _ = os.setenv("TMPDIR", self.tmpdir_back, overwrite=True)
+        _ = os.setenv("TEMP", self.temp_back, overwrite=True)
+        _ = os.setenv("TMP", self.tmp_back, overwrite=True)
+
+    fn __exit__(inout self, error: Error) raises -> Bool:
+        self.__exit__()
+        self.clean_up_function()
+        return False
+
+
+fn _clean_up_gettempdir_test() raises:
+    var dir_without_writing_access = Path() / "dir_without_writing_access"
+    if exists(dir_without_writing_access):
+        os.rmdir(dir_without_writing_access)
+    var dir_with_writing_access = Path() / "dir_with_writing_access"
+    if exists(dir_with_writing_access):
+        os.rmdir(dir_with_writing_access)
+
+
+fn _set_up_gettempdir_test(
+    dir_with_writing_access: Path, dir_without_writing_access: Path
+) raises:
+    os.mkdir(dir_with_writing_access, mode=0o700)
+    try:
+        os.mkdir(dir_without_writing_access, mode=0o100)
+    except:
+        os.rmdir(dir_with_writing_access)
+        raise Error("Failed to setup test")
+
+
+fn test_gettempdir() raises:
+    var non_existing_dir = Path() / "non_existing_dir"
+    assert_false(
+        exists(non_existing_dir),
+        "Unexpected dir" + String(non_existing_dir),
+    )
+    var dir_without_writing_access = Path() / "dir_without_writing_access"
+    var dir_with_writing_access = Path() / "dir_with_writing_access"
+    _set_up_gettempdir_test(dir_with_writing_access, dir_without_writing_access)
+
+    var tmpdir_result: Optional[String]
+    # test TEMPDIR is used first
+    with _TempEnvForTest(
+        dir_with_writing_access,
+        non_existing_dir,
+        non_existing_dir,
+        _clean_up_gettempdir_test,
+    ):
+        tmpdir_result = gettempdir()
+        assert_true(tmpdir_result, "Failed to get temporary directory")
+        assert_equal(
+            tmpdir_result.value()[],
+            String(dir_with_writing_access),
+            "expected to get:" + String(dir_with_writing_access),
+        )
+
+    # test gettempdir falls back to TEMP
+    with _TempEnvForTest(
+        non_existing_dir,
+        dir_with_writing_access,
+        non_existing_dir,
+        _clean_up_gettempdir_test,
+    ):
+        tmpdir_result = gettempdir()
+        assert_true(tmpdir_result, "Failed to get temporary directory")
+        assert_equal(
+            tmpdir_result.value()[],
+            String(dir_with_writing_access),
+            "expected to get:" + String(dir_with_writing_access),
+        )
+
+    # test gettempdir falls back to TMP
+    with _TempEnvForTest(
```

Comment:
I guess that this test is going to fail in the internal CI, similar to what happens here:
https://github.com/modularml/mojo/blob/011bf40a304078b4471fe9ca18f4101b19943aa6/stdlib/test/os/test_mkdir_and_rmdir.mojo#L84-L94